src/net/server/batcher.rs

```diff
@@ -0,0 +1,473 @@
+
+/// The result of attempting to push a resource record into a
+/// [ResourceRecordBatcher].
+pub enum PushResult<Target> {
+    /// The resource record was successfully pushed and more will fit.
+    PushedAndReadyForMore,
+
+    /// The resource record was successfully pushed and no more will fit.
+    PushedAndLimitReached(AnswerBuilder<StreamTarget<Target>>),
+
+    /// The resource record was not pushed because the message is full.
+    NotPushedMessageFull(AnswerBuilder<StreamTarget<Target>>),
+
+    /// The resource record was not pushed but trying again might work.
+    Retry,
+}
+
+//------------ ResourceRecordBatcher ------------------------------------------
+
+/// An accumulator of resource records into one or more DNS responses.
+pub trait ResourceRecordBatcher<RequestOctets, Target>
+where
+    RequestOctets: Octets,
+    Target: Composer + Default,
+{
+    /// The type returned on error.
+    type Error: From<PushError> + Debug;
+
+    /// Attempt to push a single resource record into a response message.
+    #[allow(clippy::result_unit_err)]
+    fn push(
+        &mut self,
+        record: impl ComposeRecord,
+    ) -> Result<PushResult<Target>, Self::Error>;
+
+    /// Signal that the last resource record has been pushed for this response
+    /// sequence.
+    #[allow(clippy::result_unit_err)]
+    fn finish(&mut self) -> Result<(), Self::Error>;
+
+    /// Creates a new `AnswerBuilder` for the given request message.
+    ///
+    /// The default implementation sets the RCODE of the new builder to
+    /// `NOERROR`.
+    fn mk_answer_builder(
+        &self,
+        msg: &Message<RequestOctets>,
+    ) -> Result<AnswerBuilder<StreamTarget<Target>>, PushError> {
+        let builder = mk_builder_for_target();
+        builder.start_answer(msg, Rcode::NOERROR)
+    }
+}
+
+//------------ Callbacks ------------------------------------------------------
+
+/// A set of callback functions for use with [`CallbackBatcher`].
+pub trait Callbacks<RequestOctets, Target, T>
+where
+    RequestOctets: Octets,
+    Target: Composer + Default,
+{
+    /// The type returned on error.
+    type Error: From<PushError> + Debug;
+
+    /// Prepare a message builder to push records into.
+    fn batch_started(
+        _state: &T,
+        msg: &Message<RequestOctets>,
+    ) -> Result<AnswerBuilder<StreamTarget<Target>>, PushError> {
+        let builder = mk_builder_for_target();
+        let answer = builder.start_answer(msg, Rcode::NOERROR)?;
+        Ok(answer)
+    }
+
+    /// A record has been pushed. Is the message now full?
+    ///
+    /// Return true if it is full, false if there is still space.
+    fn record_pushed(
+        _state: &T,
+        _answer: &AnswerBuilder<StreamTarget<Target>>,
+    ) -> bool {
+        false
+    }
+
+    /// Do something with the completed message.
+    #[allow(clippy::result_unit_err)]
+    fn batch_ready(
+        _state: &T,
```

Comment:
Maybe these parameters do not need an underscore?

Reply:
Fixed.